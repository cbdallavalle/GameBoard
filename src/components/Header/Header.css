header {
  position: absolute;
  top: 5px;
  right: 5px;
  width: 180px;
  display: flex;
  flex-direction: row;
  font-family: 'Righteous', cursive;
  align-items: center;
  justify-content: center;
  background-color: #ffffffe3;
  border-radius: 5px;
}

header:hover {
  cursor: pointer;
}

header button {
  height: 25px;
  border: none;
  font-size: 14px;
  font-family: 'Open Sans', sans-serif;
  background-color: unset;
  border-left: 2px solid #9b9a9a;
  font-weight: bold;
}

header button:hover {
  cursor: pointer;
}

header h4 {
  padding: 0 10px;
}