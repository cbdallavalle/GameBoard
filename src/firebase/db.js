import { db } from './firebase';

export const doCreateUser = (id, firstName, lastName, email) =>
  db.ref(`users/${id}`).set({
    user: {
      id,
      firstName,
      lastName,
      email,
    }
  });

export const onceGetUsers = async () => {
  try {
    const snapshot = await db.ref('users').once('value');
    if (snapshot.status < 300) {
      const value = await snapshot.val();
      return value
    } else {
      throw new Error('unable to fetch users')
    }
  } catch (error) {
    throw new Error('unable to fetch users')
  }
}

export const doWriteFavoriteData = async (userId, newFavorites) => {
  db.ref('users/' + userId + '/favorites').set(newFavorites);
}

export const doAddFavoriteData = async (userId, favorite) => {
  const existingFavorites = await getFavorites(userId);
  const favoriteToAdd = checkFavoriteDuplication(existingFavorites, favorite);
  doWriteLastFavoritedData(userId, favorite)
  const newFavorites = {...existingFavorites, ...favoriteToAdd }
  this.doWriteFavoriteData(userId, newFavorites)
}

export const doDeleteFavoriteData = async (userId, favorite) => {
  const existingFavorites = await getFavorites(userId);
  delete existingFavorites[favorite.name]
  this.doWriteFavoriteData(userId, existingFavorites)
}

export const doWriteLastFavoritedData = async (userId, favorite) => {
  db.ref('users/' + userId + '/lastFavorited').set(favorite);
}

export const checkFavoriteDuplication = (existingFavorites, favorite) => {
  const { name, id, description, image, thumbnail } = favorite;
  const review = "No review - click to add one";
  const favoriteToAdd = { [name]: { id, description, image, thumbnail, name, review } }
  return existingFavorites && existingFavorites[favorite.name] ? {} : favoriteToAdd
}

export const doWriteFriendsData = async (userId, friendId) => {
  try {
    const friends = await getFriends(userId);
    if (friends && !friends.includes(friendId)) {
      friends.push(friendId);
      db.ref('users/' + userId + '/friends').set(friends); 
    } else if (!friends) {
      db.ref('users/' + userId + '/friends').set([friendId]); 
    }
  } catch (error) {
    throw new Error ('unable to add friend')
  }
}

export const getFriends = async (userId) => {
  const snapshot = await db.ref('users').once('value');
  const value = await snapshot.val();
  return value[userId].friends
}

export const getFavorites = async (userId) => {
  const value = await onceGetUsers();
  return value[userId].favorites;
}

export const getFriendsFavorites = async (userId) => {
  const friends = await getFriends(userId);
  const value = await onceGetUsers();
  if(friends) {
    return friends.reduce( (accu, friendId) => {
      if(value[friendId].lastFavorited) {
        accu[value[friendId].user.firstName] = value[friendId].lastFavorited
      }
      return accu
    }, {} )  
  } else {
    return {}
  }
}

export const doWriteReviewData = async (userId, favorite, review) => {
  const { name } = favorite;
  const reviewedGame = {...favorite, review}
  db.ref('users/' + userId + '/favorites/' + name).set(reviewedGame);
  await doWriteLastFavoritedData(userId, reviewedGame)
}