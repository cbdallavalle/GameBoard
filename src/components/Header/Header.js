import React from 'react';
// import { Link } from 'react-router-dom';
import { auth } from '../../firebase';
import './Header.css';

export const Header = () => {

  const checkSignOut = () => {
    try {
      auth.doSignOut();
    } catch(error) {
      return error
      console.log(error)
    }
  }

  return (
    <header>
      <h4>GameBoard</h4>
      <button type="button" onClick={checkSignOut}>
        Sign out
      </button>
    </header>
  )
}