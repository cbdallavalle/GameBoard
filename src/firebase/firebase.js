import * as firebase from 'firebase';

const config = {
  apiKey: "AIzaSyBDU0_Pa8xdAWKmtKKfo3QkbrtR56cl024",
  authDomain: "dallavalle-gameboard.firebaseapp.com",
  databaseURL: "https://dallavalle-gameboard.firebaseio.com",
  projectId: "dallavalle-gameboard",
  storageBucket: "",
  messagingSenderId: "250758219123"
};

if(!firebase.apps.length) {
  firebase.initializeApp(config)
}

const auth = firebase.auth();

export {
  auth,
}