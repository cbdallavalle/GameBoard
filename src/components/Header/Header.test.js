/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { auth } from '../../firebase';
import { Header } from './Header';

describe("Header", () => {
  const wrapper = shallow( <Header /> )
  auth.doSignOut = jest.fn()

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('checkSignOut should call auth.doSignOut on successful sign out', () => {
    wrapper.find('button').simulate('click');
    expect(auth.doSignOut).toHaveBeenCalled()
  })

  it('checkSignOut should catch an error', () => {
    auth.doSignOut = jest.fn().mockImplementation( () => {
      throw new Error('unable to log out')
    })
    wrapper.instance().checkSignOut();

    expect(wrapper.state().error).toEqual('unable to log out');
  })
})