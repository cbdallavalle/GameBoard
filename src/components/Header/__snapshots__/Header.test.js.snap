// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <h4>
                    GameBoard
          </h4>
          <button
                    onClick={[Function]}
                    type="button"
          >
                    Sign out
          </button>
</div>,
        <p />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h4>
              GameBoard
</h4>,
            <button
              onClick={[Function]}
              type="button"
>
              Sign out
</button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "GameBoard",
            },
            "ref": null,
            "rendered": "GameBoard",
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Sign out",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": "Sign out",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "p",
      },
    ],
    "type": "header",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <h4>
                        GameBoard
            </h4>
            <button
                        onClick={[Function]}
                        type="button"
            >
                        Sign out
            </button>
</div>,
          <p />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h4>
                GameBoard
</h4>,
              <button
                onClick={[Function]}
                type="button"
>
                Sign out
</button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "GameBoard",
              },
              "ref": null,
              "rendered": "GameBoard",
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Sign out",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": "Sign out",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "p",
        },
      ],
      "type": "header",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
